sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -1056,3 +1158,615 @@ cdef class FastEuclideanPairwiseDistancesArgKmin(PairwiseDistancesArgKmin):
                     ),
                     j + Y_start,
                 )
+
+
+cdef class PairwiseDistancesRadiusNeighborhood(PairwiseDistancesReduction):
+    """Compute radius-based neighbors for two sets of vectors.
+
+    For each row-vector X[i] of the queries X, find all the indices j of
+    row-vectors in Y such that:
+
+                        dist(X[i], Y[j]) <= radius
+
+    The distance function `dist` depends on the values of the `metric`
+    and `metric_kwargs` parameters.
+
+    Parameters
+    ----------
+    datasets_pair: DatasetsPair
+        The dataset pair (X, Y) for the reduction.
+
+    chunk_size: int, default=None,
+        The number of vectors per chunk. If None (default) looks-up in
+        scikit-learn configuration for `pairwise_dist_chunk_size`,
+        and use 256 if it is not set.
+
+    n_threads: int, default=None
+        The number of OpenMP threads to use for the reduction.
+        Parallelism is done on chunks and the sharding of chunks
+        depends on the `strategy` set on
+        :meth:`~PairwiseDistancesRadiusNeighborhood.compute`.
+
+        See _openmp_effective_n_threads, for details about
+        the specification of n_threads.
+
+    radius: float
+        The radius defining the neighborhood.
+    """
+
+    cdef:
+        DTYPE_t radius
+
+        # DistanceMetric compute rank-preserving surrogate distance via rdist
+        # which are proxies necessitating less computations.
+        # We get the equivalent for the radius to be able to compare it against
+        # vectors' rank-preserving surrogate distances.
+        DTYPE_t r_radius
+
+        # Neighbors indices and distances are returned as np.ndarrays of np.ndarrays.
+        #
+        # For this implementation, we want resizable buffers which we will wrap
+        # into numpy arrays at the end. std::vector comes as a handy interface
+        # for interacting efficiently with resizable buffers.
+        #
+        # Though it is possible to access their buffer address with
+        # std::vector::data, they can't be stolen: buffers lifetime
+        # is tied to their std::vector and are deallocated when
+        # std::vectors are.
+        #
+        # To solve this, we dynamically allocate std::vectors and then
+        # encapsulate them in a StdVectorSentinel responsible for
+        # freeing them when the associated np.ndarray is freed.
+        vector[vector[ITYPE_t]] * neigh_indices
+        vector[vector[DTYPE_t]] * neigh_distances
+
+        # Used as array of pointers to private datastructures used in threads.
+        vector[vector[ITYPE_t]] ** neigh_indices_chunks
+        vector[vector[DTYPE_t]] ** neigh_distances_chunks
```

Comment:
Could this be the following?

```python
unique_ptr[vector[vector[DTYPE_t]]] neigh_distances

vector[shared_ptr[vector[vector[DTYPE_t]]]] neigh_distances_chunks
``` 

It's a shared pointer because for `X_init` `neigh_distances_chunks` does not own the pointer, but for `Y_init`, `neigh_distances_chunks` owns the pointer.

In general, I find it a little strange to mix raw pointers with c++ vectors. Although, I do not think this point blocks the PR.

Reply:
In fact, `unique_ptr` can't be used because `neigh_distances` (and respectively `neigh_indices`) referenced structures must be also referenced in `_parallel_on_X_init_chunk` by `neigh_distances_chunks` (and respectively `neigh_distances_chunks`) and `unique_ptr` can't be coerced to `shared_ptr`.

Using `shared_ptr` works though and I think it's better than using raw pointers. See 05adeb0f1f2ef78b1573c2fb2c52c2762c649b31.

Reply:
I think `unique_ptr` works since shared pointers can be constructed from unique ones. Item 13 on the [shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) docs.

I opened https://github.com/jjerphan/scikit-learn/pull/10 to try it out.